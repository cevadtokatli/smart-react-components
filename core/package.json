{
  "name": "@smart-react-components/core",
  "version": "2.0.0-beta-0",
  "license": "MIT",
  "main": "index.js",
  "description": "SRC core includes the theme, Styled properties, and Element components.",
  "keywords": [
    "javascript",
    "typescript",
    "react",
    "styled-components"
  ],
  "author": {
    "name": "Cevad Tokatlı",
    "email": "cevadtokatli@hotmail.com"
  },
  "contributors": [
    {
      "name": "Cemil Tokatlı",
      "email": "cemil@tokatli.me"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cevadtokatli/smart-react-components/tree/v2/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "yarn clear && tsc --outDir lib && yarn copy",
    "clear": "rm -rf lib",
    "copy": "cp package.json lib/package.json && cp README.md lib/README.md && cp ../LICENSE lib/LICENSE",
    "dev": "tsc -w --outDir lib",
    "lint": "eslint src",
    "release": "../scripts/release.sh core",
    "test": "jest"
  },
  "devDependencies": {
    "@types/color": "3.0.3",
    "@types/enzyme": "3.10.12",
    "@types/jest": "29.0.1",
    "@types/react": "18.0.19",
    "@types/react-dom": "18.0.6",
    "@types/styled-components": "5.1.26",
    "@wojtekmaj/enzyme-adapter-react-17": "0.6.7",
    "enzyme": "3.11.0",
    "eslint": "8.0.1",
    "jest": "29.0.3",
    "jsdom": "20.0.0",
    "jsdom-global": "3.0.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "styled-components": "5.3.5",
    "ts-jest": "29.0.0",
    "typescript": "4.8.3"
  },
  "dependencies": {
    "color": "4.2.3"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "styled-components": "^5.3.5"
  }
}
