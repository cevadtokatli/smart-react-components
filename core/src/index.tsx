export * from './theme'
export { default as createTheme } from './theme'
export * from './styled-props'
export * from './Element'
