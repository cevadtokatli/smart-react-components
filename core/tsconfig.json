{
  "extends": "../tsconfig.json",
  "include": ["src/**/*.tsx"],
  "exclude": ["node_modules", "src/**/*.test.tsx"]
}
