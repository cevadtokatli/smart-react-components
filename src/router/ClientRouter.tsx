import React from "react"
import {JSXElementProps} from "../props"
import DV from "../default-value"
import reducer, {getInitialState} from "./reducer"
import {PreloadModule, CancelCallback, RouteProps} from "../types/router"
import RouterContext from "./RouterContext"
import RouterHelper from "../helper/RouterHelper"
import {setUrl, setNewUrl} from "./actions" 
import {setPercentage, setCancelCallback} from "./actions"
import DOMHelper from "../helper/DOMHelper"
import ProgressBar from "./ProgressBar"

declare global {
    interface History {
        push: (url:string) => void
        replace: (url:string) => void
    }
}

interface Props {
    children: JSX.Element|JSX.Element[]
    params?: any
    progressBarProps?: JSXElementProps
    routes: RouteProps[]
}

const ClientRouter: React.FC<Props> = ({children,params={},progressBarProps=DV.JSX_ELEMENT_PROPS,routes}) => {
    const [state, dispatch] = React.useReducer(reducer, getInitialState())

    const setRouterUrl = React.useCallback((_url:string) => {
        (async function() {
            const url = RouterHelper.setUrl(_url)
            let key = new Date().getTime().toString()
            const setPercentage$ = (payload:number) => dispatch(setPercentage(payload, key))
            const setCancelCallback$ = (payload:CancelCallback) => dispatch(setCancelCallback(payload, key))
            let loaderMethods = []
            await RouterHelper.getLoaderMethods(routes, url.pathname, url.query, loaderMethods, params, state.url, setPercentage$, setCancelCallback$);

            if(loaderMethods.length == 0)
                dispatch(setUrl(url))
            else {
                dispatch(setNewUrl(url, key))

                try {
                    setPercentage$(10)
                    for(let i in loaderMethods) {
                        await loaderMethods[i]();
                    }
                    setPercentage$(100)
                } catch(ignored) {}
            }
        }())
    }, [state.url])

    const popstate = () => setRouterUrl(RouterHelper.getUrl())

    React.useEffect(() => {
        history.push = (url:string) => {
            history.pushState({}, null, url)
            setRouterUrl(RouterHelper.getUrl())
        }

        history.replace = (url:string) => {
            history.replaceState({}, null, url)
            setRouterUrl(RouterHelper.getUrl())
        }

        DOMHelper.addEventListener(window, ["popstate"], popstate)

        return () => {
            DOMHelper.removeEventListener(window, ["popstate"], popstate)
        }
    }, [state.url])

    return (
        <RouterContext.Provider value={{state,dispatch}}>
            <ProgressBar progressBarProps={progressBarProps} />
            {children}
        </RouterContext.Provider>
    )
}

export default ClientRouter