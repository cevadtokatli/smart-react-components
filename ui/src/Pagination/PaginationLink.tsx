import A from '@smart-react-components/core/Element/A'
import useLink, { LinkProps } from '@smart-react-components/router/hooks/useLink'
import React from 'react'
import WaveEffect from '../WaveEffect'
import { PrivateProps as PaginationItemPrivateProps } from './PaginationItem'
import PaginationItemElement from '../components/Pagination/PaginationItemElement'

export interface Props extends LinkProps {
  isActive?: boolean
}

const PaginationLink: React.FC<Props> = (props: Props & PaginationItemPrivateProps) => {
  const { children, handleOnClick, isActive, to } = useLink({
    children: props.children,
    isDisabled: props.isDisabled,
    isExact: props.isExact,
    onClick: null,
    path: props.path,
    to: props.to,
  })

  const AnchorEl = (
    <A
      {...props.elementProps}
      href={to}
      onClick={handleOnClick}
    >
      {children}
    </A>
  )
  const El = props.hasWaveEffect ? <WaveEffect palette={props.waveEffectPalette}>{AnchorEl}</WaveEffect> : AnchorEl

  return (
    <PaginationItemElement
      isActive={props.isActive ?? isActive}
      isDisabled={props.isDisabled}
      isFixedSize={props.isFixedSize}
      isOutline={props.isOutline}
      isSoft={props.isSoft}
      palette={props.palette}
      shape={props.shape}
    >
      {El}
    </PaginationItemElement>
  )
}

PaginationLink.defaultProps = {
  isExact: true,
}

export default PaginationLink
