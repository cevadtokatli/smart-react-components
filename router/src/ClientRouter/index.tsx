import { JSXChildren } from '@smart-react-components/core/types'
import { addEventListener, removeEventListener } from '@smart-react-components/core/util/dom'
import React from 'react'
import reducer, { generateInitialState, setActivatingURL, setActiveURL, setCancelCallback, setPercentage } from '../reducer'
import RouterContext from '../RouterContext'
import RoutesContext from '../RoutesContext'
import { CancelCallback, RouteModule } from '../types'
import { generateURL, getFullpath, loadModules } from '../util'

export interface Props {
  children: JSXChildren
  fallback?: JSX.Element
  params?: any
  progressBar?: JSX.Element
  routes: RouteModule[]
}

export const modules: object = {}

const ClientRouter: React.FC<Props> = ({ children, fallback, params, routes, progressBar }) => {
  const [state, dispatch] = React.useReducer(reducer, generateInitialState())

  const handleURLChange = () => {
    const url = generateURL(getFullpath())

    if (
      state.activatingURL?.fullpath === url.fullpath
      || (!state.activatingURL && state.activeURL.fullpath === url.fullpath)
    ) {
      return
    }

    (async function () {
      const modulesToInvokeGetMethods = await loadModules(state.activeURL, url, routes, modules)

      if (modulesToInvokeGetMethods.length === 0) {
        dispatch(setActiveURL(url))
      } else {
        const key = new Date().getTime()
        dispatch(setActivatingURL({ key, url }))
        const _setPercentage = (value: number) => dispatch(setPercentage({ key, value }))
        const _setCancelCallback = (callback: CancelCallback) => dispatch(setCancelCallback({ callback, key }))

        _setPercentage(10)

        for (const i in modulesToInvokeGetMethods) {
          await modules[modulesToInvokeGetMethods[i].module as any].get?.(
            modulesToInvokeGetMethods[i].match,
            url,
            _setPercentage,
            _setCancelCallback,
            params,
          )
        }

        _setPercentage(100)
      }
    }())
  }

  const redirect = React.useCallback((to, isNewPage) => {
    if (isNewPage) {
      window.open(to, '_blank')
    } else {
      window.location.href = to
    }
  }, [])

  const push = React.useCallback((to: string) => {
    history.pushState({}, null, to)
    handleURLChange()
  }, [state.activeURL.fullpath, state.activatingURL?.fullpath])

  const replace = React.useCallback((to: string) => {
    history.replaceState({}, null, to)
    handleURLChange()
  }, [state.activeURL.fullpath, state.activatingURL?.fullpath])

  React.useEffect(() => {
    addEventListener(window, ['popstate'], handleURLChange)

    return () => {
      removeEventListener(window, ['popstate'], handleURLChange)
    }
  }, [state.activeURL.fullpath, state.activatingURL?.fullpath])

  return (
    <RouterContext.Provider value={{ fallback, state, dispatch, modules, push, redirect, replace }}>
      <RoutesContext.Provider value={routes}>
        { progressBar && React.cloneElement(progressBar, { value: state.percentage }) }
        { children }
      </RoutesContext.Provider>
    </RouterContext.Provider>
  )
}

export default ClientRouter
