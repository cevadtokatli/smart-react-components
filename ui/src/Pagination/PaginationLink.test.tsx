import { wrapTheme } from '@smart-react-components/core/test'
import ClientRouter from '@smart-react-components/router/ClientRouter'
import { render } from '@testing-library/react'
import React from 'react'
import PaginationLink from './PaginationLink'

describe('<PaginationLink />', () => {
  let props

  beforeAll(() => {
    props = {
      hasWaveEffect: true,
      isFixedSize: false,
      isOutline: false,
      isSoft: false,
      palette: 'primary',
      shape: 'rectangle',
      to: '/',
      waveEffectPalette: 'light',
    }
  })

  it('should render component', () => {
    const { asFragment } = render(wrapTheme(
      <ClientRouter routes={[]}>
        <PaginationLink {...props}>1</PaginationLink>
      </ClientRouter>
    ))
    expect(asFragment()).toMatchSnapshot()
  })
})
