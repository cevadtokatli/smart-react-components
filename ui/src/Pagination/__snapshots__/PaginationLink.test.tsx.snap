// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationLink /> should render component 1`] = `
<DocumentFragment>
  .c1 {
  isolation: isolate;
  overflow: hidden;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c0 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: .2525rem .2525rem;
}

.c0 > a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: solid 1px;
  border-radius: .125rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 300ms 0s ease-in-out;
  transition: 300ms 0s ease-in-out;
  -webkit-transition-property: background,border,color,fill;
  transition-property: background,border,color,fill;
  padding: .4rem .65rem;
  background: #0556B2;
  border-color: #0552A9;
  color: #fff;
  fill: #fff;
}

<li
    class=" c0"
    shape="rectangle"
  >
    <a
      class="c1"
      href="/"
    >
      1
    </a>
  </li>
</DocumentFragment>
`;
